raft/src/main/java/org/apache/kafka/raft/FileBasedStateStore.java

```diff
@@ -0,0 +1,164 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.raft;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.fasterxml.jackson.databind.node.ShortNode;
+import org.apache.kafka.common.utils.Utils;
+import org.apache.kafka.raft.generated.QuorumStateData;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.EOFException;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.util.OptionalInt;
+
+/**
+ * Local file based quorum state store. It takes the JSON format of {@link QuorumStateData}
+ * with an extra data version number as part of the data for easy deserialization.
+ *
+ * Example format:
+ * <pre>
+ * {"clusterId":"",
+ *   "leaderId":1,
+ *   "leaderEpoch":2,
+ *   "votedId":-1,
+ *   "appliedOffset":0,
+ *   "currentVoters":[],
+ *   "targetVoters":[],
+ *   "data_version":0}
+ * </pre>
+ * */
+public class FileBasedStateStore implements QuorumStateStore {
+    private static final Logger log = LoggerFactory.getLogger(FileBasedStateStore.class);
+
+    private final File stateFile;
+
+    private final QuorumStateData cachedState;
+
+    static final String DATA_VERSION = "data_version";
+
+    public FileBasedStateStore(final File stateFile, final ReplicatedLog metadataLog) throws IOException {
+        this.stateFile = stateFile;
+        cachedState = loadState(stateFile, metadataLog);
+        log.info("Loaded quorum state from local: {}", cachedState);
+    }
+
+    private QuorumStateData loadState(File file, final ReplicatedLog metadataLog) throws IOException {
+        try (final BufferedReader reader = Files.newBufferedReader(file.toPath())) {
+            final ObjectMapper objectMapper = new ObjectMapper();
+            final ObjectNode dataObject = (ObjectNode) objectMapper.readTree(readLine(reader));
+            final short dataVersion = dataObject.get(DATA_VERSION).shortValue();
+
+            QuorumStateData data = new QuorumStateData();
+            data.fromJson(dataObject, dataVersion);
+            return data;
+        } catch (final NoSuchFileException e) {
+            return recoverStateFromLog(file, metadataLog);
+        } catch (final EOFException e) {
+            delete();
```

Comment:
I'm wondering in general about the policy of deleting the file after seeing corrupt data. It might be the simplest option. An alternative is to raise an error to the user. If we think deletion is generally safe (I think I do), then maybe we can save the user the trouble and recreate it ourselves. With that said, we should probably handle all causes of corruption consistently. So if the JSON doesn't parse or if the parsed result is not an object, then maybe we should also delete it.